sdk/Pulumi/Provider/PropertyValueSerializer.cs

```diff
@@ -0,0 +1,751 @@
+        }
+
+        public async Task<PropertyValue> Serialize(object? value)
+        {
+            if (value == null)
+            {
+                return PropertyValue.Null;
+            }
+
+            var targetType = value.GetType();
+            switch (value)
+            {
+                case string stringValue:
+                    return new PropertyValue(stringValue);
+                case int integer:
+                    return new PropertyValue(integer);
+                case bool boolean:
+                    return new PropertyValue(boolean);
+                case double number:
+                    return new PropertyValue(number);
+                case Asset asset:
+                    return new PropertyValue(asset);
+                case Archive archive:
+                    return new PropertyValue(archive);
+                case PropertyValue propertyValue:
+                    return propertyValue;
+            }
+
+            if (targetType.IsGenericType && targetType.GetGenericTypeDefinition() == typeof(Nullable<>))
+            {
+                var valueField = targetType.GetField("Value")?.GetValue(value);
+                return await Serialize(valueField);
+            }
+
+            if (targetType.IsArray)
+            {
+                var elements = ImmutableArray.CreateBuilder<PropertyValue>();
+                foreach (var element in (Array)value)
+                {
+                    var item = await Serialize(element);
+                    elements.Add(item);
+                }
+
+                return new PropertyValue(elements.ToImmutableArray());
+            }
+
+            if (value is IList list)
+            {
+                var elements = ImmutableArray.CreateBuilder<PropertyValue>();
+
+                if (Serializer.InitializedByDefault(list))
+                {
+                    return new PropertyValue(elements.ToImmutableArray());
+                }
+
+                foreach (var element in list)
+                {
+                    var item = await Serialize(element);
+                    elements.Add(item);
+                }
+
+                return new PropertyValue(elements.ToImmutableArray());
+            }
+
+            if (value is IDictionary dictionary)
+            {
+                var propertyObject = ImmutableDictionary.CreateBuilder<string, PropertyValue>();
+                foreach (var pair in dictionary)
+                {
+                    if (pair is DictionaryEntry keyValuePair && keyValuePair.Key is string key)
+                    {
+                        var serializedValue = await Serialize(keyValuePair.Value);
+                        propertyObject.Add(key, serializedValue);
+                    }
+                }
+
+                return new PropertyValue(propertyObject.ToImmutableDictionary());
+            }
+
+            if (targetType.IsEnum)
+            {
+                var enumValue = Enum.GetName(targetType, value);
```

Comment:
We write these as their names rather than values? That feels a bit surprising, again I'd copy the behavior from System.Text.Json for this and AFAIK by default it writes values.

Reply:
Changed implementation so that enums write out their values instead names ✅ will need to remember that when doing schema-gen